plugins/woocommerce/src/Internal/DataStores/Orders/LegacyDataCleanup.php

```diff
@@ -0,0 +1,320 @@
+	 *
+	 * @var BatchProcessingController
+	 */
+	private $batch_processing;
+
+	/**
+	 * The legacy handler to use for the actual cleanup.
+	 *
+	 * @var LegacyHandler
+	 */
+	private $legacy_handler;
+
+	/**
+	 * The data synchronizer object to use.
+	 *
+	 * @var DataSynchronizer
+	 */
+	private $data_synchronizer;
+
+	/**
+	 * Logger object to be used to log events.
+	 *
+	 * @var \WC_Logger
+	 */
+	private $error_logger;
+
+	/**
+	 * Class constructor.
+	 */
+	public function __construct() {
+		self::add_filter( 'pre_update_option_' . self::OPTION_NAME, array( $this, 'pre_update_option' ), 999, 2 );
+		self::add_action( 'add_option_' . self::OPTION_NAME, array( $this, 'process_added_option' ), 999, 2 );
+		self::add_action( 'update_option_' . self::OPTION_NAME, array( $this, 'process_updated_option' ), 999, 2 );
+		self::add_action( 'delete_option_' . self::OPTION_NAME, array( $this, 'process_deleted_option' ), 999 );
+		self::add_action( 'shutdown', array( $this, 'maybe_reset_state' ) );
+	}
+
+	/**
+	 * Class initialization, invoked by the DI container.
+	 *
+	 * @param BatchProcessingController $batch_processing  The batch processing controller to use.
+	 * @param LegacyDataHandler         $legacy_handler    Legacy order data handler instance.
+	 * @param DataSynchronizer          $data_synchronizer Data synchronizer instance.
+	 * @internal
+	 */
+	final public function init( BatchProcessingController $batch_processing, LegacyDataHandler $legacy_handler, DataSynchronizer $data_synchronizer ) {
+		$this->legacy_handler    = $legacy_handler;
+		$this->data_synchronizer = $data_synchronizer;
+		$this->batch_processing  = $batch_processing;
+		$this->error_logger      = wc_get_logger();
+	}
+
+	/**
+	 * A user friendly name for this process.
+	 *
+	 * @return string Name of the process.
+	 */
+	public function get_name(): string {
+		return 'Order legacy data cleanup';
+	}
+
+	/**
+	 * A user friendly description for this process.
+	 *
+	 * @return string Description.
+	 */
+	public function get_description(): string {
+		return 'Cleans up order data from legacy tables.';
+	}
+
+	/**
+	 * Get total number of pending records that require update.
+	 *
+	 * @return int Number of pending records.
+	 */
+	public function get_total_pending_count(): int {
+		return $this->legacy_handler->count_orders_for_cleanup();
+	}
+
+	/**
+	 * Returns the batch with records that needs to be processed for a given size.
+	 *
+	 * @param int $size Size of the batch.
+	 * @return array Batch of records.
+	 */
+	public function get_next_batch_to_process( int $size ): array {
+		return array_map(
+			'absint',
+			$this->legacy_handler->get_orders_for_cleanup( array(), $size )
+		);
+	}
+
+	/**
+	 * Process data for current batch.
+	 *
+	 * @param array $batch Batch details.
+	 */
+	public function process_batch( array $batch ): void {
+		// This is a destructive operation, so check if we need to bail out just in case.
+		if ( ! $this->should_run() ) {
```

Comment:
Maybe this check should also be added in `get_next_batch_to_process` and `get_total_pending_count` so that they both return 0 when we need to bail.

Reply:
Done in ad534451f8f5ed14e04a373b4abe6f0691b98182.